lib/_http_parser.js

```diff
@@ -0,0 +1,890 @@
+            IP-literal      = "[" ( IPv6address | IPvFuture ) "]"
+              IPv6address   =                               6( h16 ":" ) ls32
+                               |                       "::" 5( h16 ":" ) ls32
+                               | [               h16 ] "::" 4( h16 ":" ) ls32
+                               | [ *1( h16 ":" ) h16 ] "::" 3( h16 ":" ) ls32
+                               | [ *2( h16 ":" ) h16 ] "::" 2( h16 ":" ) ls32
+                               | [ *3( h16 ":" ) h16 ] "::"    h16 ":"   ls32
+                               | [ *4( h16 ":" ) h16 ] "::"              ls32
+                               | [ *5( h16 ":" ) h16 ] "::"              h16
+                               | [ *6( h16 ":" ) h16 ] "::"
+                h16         = 1*4HEXDIG
+                ls32        = ( h16 ":" h16 ) | IPv4address
+                IPv4address = dec-octet "." dec-octet "." dec-octet "."
+                              dec-octet
+                dec-octet   =   DIGIT                 ; 0-9
+                              | %x31-39 DIGIT         ; 10-99
+                              | "1" 2DIGIT            ; 100-199
+                              | "2" %x30-34 DIGIT     ; 200-249
+                              | "25" %x30-35          ; 250-255
+            reg-name        = *( unreserved | pct-encoded | sub-delims )
+          port              = *DIGIT
+        path-abempty        = *( "/" segment )
+        path-absolute       = "/" [ segment-nz *( "/" segment ) ]
+          segment-nz        = 1*pchar
+        path-rootless       = segment-nz *( "/" segment )
+        path-empty          = 0<pchar>
+  authority-form            = authority
+  asterisk-form             = "*"
+*/
+
+// Note: AT LEAST a space is technically required after the status code, but
+// joyent/http-parser allows a CRLF immediately following the status code, so we
+// do also for backwards compatibility ...
+var RE_STATUS_LINE = /^HTTP\/1\.([01]) ([0-9]{3})(?: (.*))?$/;
+
+var RE_HEADER = /^([!#$%'*+\-.^_`|~0-9A-Za-z]+):(.*)$/;
+
+var STATE_REQ_LINE = 0;
+var STATE_STATUS_LINE = 1;
+var STATE_HEADER = 2;
+var STATE_BODY_LITERAL = 3;
+var STATE_BODY_EOF = 4;
+var STATE_BODY_CHUNKED_SIZE = 5;
+var STATE_BODY_CHUNKED_BYTES = 6;
+var STATE_BODY_CHUNKED_BYTES_CRLF = 7;
+var STATE_COMPLETE = 8;
+
+var FLAG_CHUNKED = 1 << 0;
+var FLAG_CONNECTION_KEEP_ALIVE = 1 << 1;
+var FLAG_CONNECTION_CLOSE = 1 << 2;
+var FLAG_CONNECTION_UPGRADE = 1 << 3;
+var FLAG_TRAILING = 1 << 4;
+var FLAG_UPGRADE = 1 << 5;
+var FLAG_SKIPBODY = 1 << 6;
+var FLAG_SHOULD_KEEP_ALIVE = 1 << 7;
+var FLAG_ANY_UPGRADE = FLAG_UPGRADE | FLAG_CONNECTION_UPGRADE;
+
+function HTTPParser(type) {
+  this.onHeaders = undefined;
+  this.onBody = undefined;
+  this.onComplete = undefined;
+
+  // extra stuff tagged onto parser object by core modules/files
+  this.onIncoming = undefined;
+  this.incoming = undefined;
+  this.socket = undefined;
```

Comment:
I prefer `null`, since it's explicitly, intentionally generated by the programmer, vs. `undefined` can be generated by language operations (`MemberExpression`s) unintentionally.